src/index.ts

```diff
@@ -82,6 +95,69 @@ function tryParseFloat(value: string) {
 	return result;
 }
 
+function normalizeNumber(value: string, allowDot: boolean = true): string {
+	const match = numberPrefixPattern.exec(value);
+	if (!match) {
+		throw new Error('Invalid number: ' + value);
+	}
+	const [, sign, dot] = match;
+	let dots = dot ? 1 : 0;
+	const endingZero = /0$/.test(value);
+	if (sign === '+') {
+		value = value.substring(1);
+	}
+	if (dot === '.') {
+		if (!allowDot) {
+			throw new Error('Invalid number (too many dots): ' + value);
+		}
+		if (sign === '-') {
+			value = '-0' + value.substring(1);
+		} else {
+			value = '0' + value;
+		}
+	} else if (endingZero || !allowDot) {
+		dots = countDots(value);
+	}
+	if (dots > 0) {
+			if (!allowDot) {
+				throw new Error('Invalid number (too many dots): ' + value);
+			}
+			if (endingZero) {
+				value = value.replace(/\.?0+$/, '');
+			}
+	}
+	return value;
+}
+
+function tryParseStrict(value: string): number {
+	const nval = normalizeNumber(value);
+	const result = parseFloat(nval);
+	if (result.toString() !== nval) {
+		if (verifyZero(value) || verifyENotation(value)) {
+			return result;
+		}
+		throw new Error('Invalid number: ' + value);
+	}
+	return result;
+}
+
+function verifyZero(value: string) {
+	return /^[-+]?0\.0+$/.test(value);
```

Comment:
The following works w/o breaking tests:
```ts
return parseFloat(value) === 0;
```

Reply:
I guess so, but `parseFloat()` will always ignore trailing non-number parts in the string -- I have not thought too hard about this (in this instance), but it may be that there is such a case that it would fail here (i.e. falsely claim a non-valid input as number)

Reply:
Yes, this is the discussion I was trying to spark here. Those other cases might be worth testing. You could also consider `Math.abs(+value) === 0`.